### ISSUE TITLE
Bread-first search

### ISSUE THREAD

**Issue by stefnotch** [mgatelabs/GenshinSolvers#1]
The fact that this solver is [calling a random number generator](https://github.com/mgatelabs/GenshinSolvers/blob/6224c4466bf97ed9a62610d5d0a32f5a5953e405/src/SpinPuzzle.java#L38) makes me think that it simply tries out stuff until something works. Is that assessment correct?

An even more ~~overblown~~ well engineered approach would be to try out all possibilities in a breadth-first manner. In fact, I did write something like this. And since nobody said anything about being reasonable, I wrote it in Haskell.

Pro:
It's super short and works beautifully

Con:
It's in Haskell. Not sure if that's better or worse than Java.

```haskell
type Cube = Int
type Step = Int
type PossibleRotation = [Int] -- As long as the cube array
data State = S [Step] [Cube]

solvePuzzle :: [State] -> [PossibleRotation] -> [Step]
solvePuzzle states possibleRotations
  | null solutions = solvePuzzle (concatMap (\s -> zipWith (applyRotation s) possibleRotations [0..]) states) possibleRotations
  | otherwise = getSteps (head solutions)
  where solutions = filter (\(S _ cubes) -> all (==0) cubes) states

getSteps :: State -> [Step]
getSteps (S steps _) = steps

applyRotation :: State -> PossibleRotation -> Int -> State
applyRotation (S steps cubes) rotation rotationId = S (steps++[rotationId]) (zipWith (\cube rot -> (cube + rot) `mod` 4) cubes rotation)

-- solvePuzzle [(S [] [0,2,0,0])] [[1,1,0,1],[1,1,1,0],[0,1,1,1],[1,0,1,1]]
-- solvePuzzle [(S [] [0,2,0,0])] [[1,1,0,1],[1,1,1,0],[0,1,1,1],[1,0,1,1]]
```

Feel free to use the code. Though, if one wants to make something actually usable, I'd recommend writing it in Typescript and having a proper, static website. I'd totally be willing to help if you're interested.

**Comment by mgatelabs** [comment]
Yes, I did it dirty and used the random number gen instead of "the smart way".  Now I just need to figure out how to read this.

**Comment by mgatelabs** [comment]
I've played and refactored a bunch.  Few setup calls and it finds a faster solution.

```
9670 Solutions found after 2294145 attempts
--Best Solution Found--
Initial Cube State
[N|N|N|E|W]
Hits
Hit #1 - Cube Index [2] - Triggers [0,4]
[E|N|E|E|N]
Hit #2 - Cube Index [2] - Triggers [0,4]
[S|N|S|E|E]
Hit #3 - Cube Index [1] - Triggers [0,2]
[W|E|W|E|E]
Hit #4 - Cube Index [0] - Triggers [2]
[N|E|N|E|E]
Hit #5 - Cube Index [0] - Triggers [2]
[E|E|E|E|E]

Color Key
Unmodified Cube
Cube to Touch
Modified Because of Touch
```

**Comment by stefnotch** [comment]
> Yes, I did it dirty and used the random number gen instead of "the smart way". Now I just need to figure out how to read this.

I can rewrite it in mostly any language you like. And actually add comments.
Here it is in Javascript. 

```js
/**
 * A possibility that we'll explore in the breadth-first search
 */
class State {
  /**
   * Which cubes did we touch to get to this state?
   */
  steps = [];
  /**
   * How are the cubes arranged? I'm using numbers (0,1,2,3)
   */
  cubes = [];

  constructor(steps, cubes) {
    this.steps = steps;
    this.cubes = cubes;
  }
}

/**
 * Applies a rotation to some cubes. A rotation just describes which cubes get rotated.
 * It then returns a new array of cubes.
 */
function applyRotation(rotation, cubes) {
  if (rotation.length != cubes.length)
    throw new Error("There has to be one rotation for every cube.");
  let newCubes = cubes.slice(); // Copy the array
  for (let i = 0; i < rotation.length; i++) {
    newCubes[i] = mod(cubes[i] + rotation[i], 4);
  }
  return newCubes;
}

function applyAllRotations(possibleRotations, state) {
  // We take every single rotation, and create a new state
  return possibleRotations.map(
    (rotation, i) =>
      new State(state.steps.concat(i), applyRotation(rotation, state.cubes))
  );
}

/**
 * states: We explore all possibilities at the same time. Could be improved by only visiting actually new states.
 * possibleRotations: If we touch cube n, which cubes get rotated
 */
function solvePuzzle(states, possibleRotations) {
  // Solutions are those possibilities where every cube is 0 (facing in the desired direction, could be made better)
  let solutions = states.filter((state) => state.cubes.every((cube) => cube == 0));

  if (solutions.length == 0) {
    // Keep searching
    const newStates = states
      .map((state) => applyAllRotations(possibleRotations, state))
      .flat();
    return solvePuzzle(newStates, possibleRotations);
  } else {
    // Yay, we found a solution
    return solutions[0].steps;
  }
}

function mod(a, b) {
  // Most languages don't implement a proper modulo that handles negative numbers
  // Instead they implement a division-remainder. Ew.
  return ((a % b) + b) % b;
}

solvePuzzle(
  // We haven't done anything yet
  // And the cube at index 1 has been rotated 2 times
  [new State([], [0, 2, 0, 0])],

  // And if we hit the cube at index N, it will get rotated accordingly
  [
    [1, 1, 0, 1],
    [1, 1, 1, 0],
    [0, 1, 1, 1],
    [1, 0, 1, 1],
  ],
);
```



> And the whole "Good" direction thing is not needed for most, except for puzzles that have a fixed direction.

I think the easiest way is to have a function that checks if a solution is valid. That way, you could easily switch out which variant to use at runtime.

> 9670 Solutions found after 2294145 attempts

Yeah, the max depth variant works. But doing it breadth-first, where you first check all solutions with a max depth of 0, then with a max depth of 1, then with a max depth of 2, and so on is neater. It guarantees that the first solution you find is also the best one.



> I did some work, think it is better now, and it doesn't check everything
> 
> [Sample](https://imgur.com/a/HToS21n)

Very nice indeed!

**Comment by stefnotch** [comment]
Oh lol, I also have a super simplistic one that assumes cubes will only ever rotate in one direction

![image](https://cdn.discordapp.com/attachments/909407955747745793/927977035761070110/unknown.png)

**Comment by mgatelabs** [comment]
I keep thinking, need to add drag and drop somewhere, or pre-program all the puzzle configurations, but if you just showed up to the puzzle, the pieces are in a fixed direction, here is the answer.

**Comment by mgatelabs** [comment]
> Oh lol, I also have a super simplistic one that assumes cubes will only ever rotate in one direction
> 
> ![image](https://camo.githubusercontent.com/36e3fdfb3335eaf8b1332aab5a0f903493f982d3089531ab5423a3eb5feaedd8/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f6174746163686d656e74732f3930393430373935353734373734353739332f3932373937373033353736313037303131302f756e6b6e6f776e2e706e67)

This puzzle is why I wrote it.

**Comment by stefnotch** [comment]
I think the easiest option is to code in every puzzle in the game. Writing something that lets you configure a puzzle seems like it'd be more trouble than it's worth.

Which makes me wonder: Is there a list of all puzzles somewhere?

**Comment by mgatelabs** [comment]
> 

On the official map it has every cube puzzle listed.  But I already solved them, so I can't play with them again to get the X triggers Y logic and initial diection.

**Comment by mgatelabs** [comment]
But I could easily change up the code and have it start at a lower max depth, and then on failure, increase it until it works / max.

**Comment by stefnotch** [comment]
I suppose there are a few reasonable options:
- Creating a new Genshin account and speedrunning to Inazuma
- Crowdsourcing it. As long as we have one puzzle and a nice UI, it should make for a decent Reddit post. (Karma go brrr.) Or we could ask on a Genshin Discord server.
- Or we could join random people and solve their puzzles
- Or looking up videos of some of the puzzles might also work

Also, how do you feel about turning this into a website? As in, boring old JavaScript, HTML, CSS, a bunch of images and hosting it on Github Pages. Costs nothing, is easy to make, and saves people the trouble of downloading something.
Of course, if you want to try out anything something different, I'm up for it. (Like Rust compiled to WebAssembly?)

**Comment by mgatelabs** [comment]
It would not be that hard to make simple js page version, since it would just be all client side.  Use a worker.  I thought about making a simple site, but I forgot about github pages, then it could have a crude map of all the puzzles to click to see the root solution and if you messed with it, enter the current facing.

I still have one of my kids accounts that hasn't don't anything yet in inazuma.

**Comment by stefnotch** [comment]
> It would not be that hard to make simple js page version, since it would just be all client side. Use a worker. I thought about making a simple site, but I forgot about github pages, then it could have a crude map of all the puzzles to click to see the root solution and if you messed with it, enter the current facing.

Yep, fully agreed. It's always nice when a web-development project boils down to client side Javascript. No complicated servers involved, haha.
And yes, the crude map is definitely a good idea.


> I still have one of my kids accounts that hasn't don't anything yet in inazuma.

Sweet! Then that beautifully solves that.

**Comment by mgatelabs** [comment]
Just need to worry about copyrights, Jenshin Impact is always pretty memeable.

**Comment by stefnotch** [comment]
![image](https://user-images.githubusercontent.com/10220080/155351846-777b53e0-726d-4f69-a678-8ceca7c7981d.png)
On the topic of Jenshin Impact, this is what my better Genshin desktop icon looks like.

**Comment by mgatelabs** [comment]
> Bwahaha yes indeed. I also propose drawing the Inazuma map instead of using the official, probably copyrighted one. Bonus points for drawing it with Microsoft Paint.

This is a task I need to give to my oldest kid, they "think they can draw".  And they will add in random sea monsters.

**Comment by mgatelabs** [comment]
Soooooo, think I found how to make a zoomable map that can use a custom image.

https://leafletjs.com/examples/crs-simple/crs-simple.html

**Comment by mgatelabs** [comment]
Did some cursory research, seems like I could do a angular project, have it build into /docs and it will present.  Don't need to go overboard, two views (world map), and puzzle.  So there could be links to specific puzzles.

**Comment by mgatelabs** [comment]
I need to eventually convert the 20 year old monolith to Angular, so seems like it will be Angular.  Should have a demo site that does nothing soon.

**Comment by stefnotch** [comment]
> I need to eventually convert the 20 year old monolith to Angular, so seems like it will be Angular. Should have a demo site that does nothing soon.

Then Angular it is! Always nice when a fun project has the side-effect of "learning something useful".
I'll look forward to the demo site! 

While I'm at it, should I do all my contributions through Github pull requests or would you prefer giving me access to the repository? Either option is equally fine with me.

**Comment by mgatelabs** [comment]
Ok, the demo should appear here https://mgatelabs.github.io/GenshinSolvers/ It should take a few minutes for it to switch.

Pull requests would be the best, I should setup a pages-dev branch for me to work under and pull them in also.

**Comment by stefnotch** [comment]
Yeah, base url stuff.
By the way, there is a slightly more elegant way of doing the Github pages stuff that doesn't require you to mix auto-generated files with the other files.

Basically
1. Move the stuff from https://github.com/mgatelabs/GenshinSolvers/tree/Pages to the `main` branch
2. Add the `docs` or whatever folder to the `.gitignore`
3. `npm install gh-pages`
4. Build an Angular project
5. And then use the `gh-pages` to automatically take everything from the `docs` folder and force-push it to the `gh-pages` branch. [I got a script for it here](https://github.com/stefnotch/vite-pseudo-template/blob/main/utils/publish.js)
6. Tell Github to use the `gh-pages` branch

**Comment by mgatelabs** [comment]
Ok, had to undo and re-do the changes, but it is now doing the magic.

But the Pages branch should be the page specific branch.  The original java app should stay in the master branch.

**Comment by stefnotch** [comment]
> 

Sounds reasonable :)
The easiest way of redoing the changes is to checkout the master branch before the revert. Then select everything except for the `.git` folder and copy it somewhere else. Then checkout the `Pages` branch. And replace everything there with what you copied previously.

**Comment by mgatelabs** [comment]
Ok, time to get serious, need to make a development branch where my changes will go, and then start merging them instead of just pushing to Pages.

And probably need to start making issues about the missing things

**Comment by stefnotch** [comment]
> Ok, time to get serious, need to make a development branch where my changes will go, and then start merging them instead of just pushing to Pages.
> 
> And probably need to start making issues about the missing things

Awesome! From the quality of life tweaks, could you please add in the "format on save" stuff? And you'll definitely want to add the `.nojekyll` file, otherwise Github pages occasionally acts up.
https://github.com/mgatelabs/GenshinSolvers/pull/6/commits/6c4f39419799f66287d5fadea990059cf72024d7

**Comment by stefnotch** [comment]
Nice, thank you :)
As soon as you got issues for the various features in place, feel free to assign me to some. Or I'll simply comment on the ones that I'd like to take care of.

**Comment by mgatelabs** [comment]
I just noticed, "bread" first search.

**Comment by stefnotch** [comment]
Nice, spotted the pun. Yes indeed, searching for a nice loaf, preferably a [cat loaf](https://en.wikipedia.org/wiki/Catloaf) should be the first thing any good algorithm does. 😋

**Comment by mgatelabs** [comment]
The demo has been updated to the latest dev build https://mgatelabs.github.io/GenshinSolvers/

- It has the Inazuma map
- It has teleport, puzzle and seven icons

It does nothing yet

**Comment by mgatelabs** [comment]
So now, is the hard part, research, need to visit a few of the puzzles, take a screen shot with the camera, remove the character.  Save the initial orientations, see what each hit does.

Then need to start building out the puzzle json, the 2nd screen for the puzzle info and solver.

**Comment by stefnotch** [comment]
As for collecting data, I think the safest option is to record it all. Fire up OBS, start recording and take notes of what each hit does.
That way, if you mess up, you still have a screen recording to fall back to.

And about the "getting to the puzzles part", I'd be up anytime to play Genshin. XD

**Comment by stefnotch** [comment]
Regarding the presentation, one thing that would be useful would be "which direction is the player looking in". As in, the direction the arrow is pointing in on the minimap.

![image](https://user-images.githubusercontent.com/10220080/155557397-65ba45fc-eebb-4a98-8f6a-a7180aa93fb9.png)

**Comment by mgatelabs** [comment]
True, which way you are facing is important.  All the small details.

**Comment by mgatelabs** [comment]
So thinking, Bootstrap to handle mobile / desktop?

**Comment by stefnotch** [comment]
Nice progress!

> True, which way you are facing is important. All the small details. But if possible we need to face true direction, so all the NESW makes sense. And some of the images may need to be edited, i'm thinking of a few puzzles where some of the cubes are a bit away.

Yeah, true. There is at least one puzzle where the pieces are separated by an ocean.


> I was thinking about threejs, i've used it in the past. It gets tricky, but could make a 3d model with an arrow facing and place it over the image, it just gets tricky to determine the orientation and position, so it scales well.
> 

I was more thinking of recreating the cubes using Three.js. Not sure how practical that would be.
One alternative, along those lines, would be using isometric pixel art. Like this
![image](https://user-images.githubusercontent.com/10220080/155572331-9e5e276c-83b7-4adf-9c0b-63a07933a4df.png)


> And at some point need to add in the translation stuff, but don't even have a solver yet, so it can wait. I don't think this is going to be very "wordy" app.

Yes, you're definitely right about that. Depending on how things go, we might be able to get away with practically zero text.


> So thinking, Bootstrap to handle mobile / desktop? Grids, fancy fields.

Sure, Bootstrap is a solid choice. Probably an overkill for this, but why not.

**Comment by mgatelabs** [comment]
I like the iso metric art style, kind of feels like minecraft, but I fear there may be a disconnect between the iso view and the real world.  Unless we specifically take the images in a certain way.  I've written a game in three js before, I can make it do things.

**Comment by stefnotch** [comment]
> I like the iso metric art style, kind of feels like minecraft, but I fear there may be a disconnect between the iso view and the real world. Unless we specifically take the images in a certain way. I've written a game in three js before, I can make it do things.

True, it might make the tool not quite as easy to use as it could be.
If we're going with the 3D Three.js approach, I wonder if stitching some Genshin screenshots together to make an environment map would work. Or if that would be too tricky to be worthwhile.

**Comment by mgatelabs** [comment]
I think fixed 3D, like the old Grim Fandango, place the 3d cubes on top of the image.  So the player just need to stand in a similar spot and shoot arrows at the cubes in the order we present.  Otherwise we would need to break out the Nvidia image tech where it makes 360 images in games, but I don't think GI does that.  Now we could add in multiple images and allow switching for some of the less simple ones, just adjust the "world" position so it all makes sense.

**Comment by stefnotch** [comment]
I agree, this sounds like the best option so far. Let's go with that then.

**Comment by stefnotch** [comment]
One quick question regarding puzzles: Do they really only ever rotate in one direction/increment the glowing leaves by 1?
If yes, then we can use the super simplistic code from https://github.com/mgatelabs/GenshinSolvers/issues/1#issuecomment-1048511959

**Comment by mgatelabs** [comment]
I'm pretty sure they all turn the same way, and they all move by 1 state, but until it has been evaluated, not 100% sure.

**Comment by stefnotch** [comment]
Here is an untested solver, it works just like the simplistic one. The main differences are that it's properly typed, doesn't do useless work and has comments.

Please do let me know when and where I should integrate it ;)

```ts
type PuzzleState = readonly number[];

type Puzzle = {
  /**
   * We represent the state with simple numbers
   */
  initialState: PuzzleState;

  /**
   * What state do we want to reach?
   */
  finalState: PuzzleState;

  /**
   * What's the maximum state (exclusive), before wrapping around.
   * 4 in the case of a rotating cube puzzle.
   */
  maximumNumber: number;

  /**
   * A state transition basically says "if we hit cube X then Y will be affected"
   */
  stateTransitions: number[][];
};

type PuzzleSolveStep = {
  previousStep: PuzzleSolveStep | null;
  touchedCube: number | null;
  state: PuzzleState;
};

/**
 * @param puzzle The puzzle to be solved
 * @returns A solution if one was found, otherwise null
 */
function solvePuzzle(puzzle: Puzzle) {
  // To prevent doing work more often than we have to
  const solvedStates = new Set<string>();

  // Which puzzle states can we reach and have not solved yet.
  const workQueue: PuzzleSolveStep[] = [
    {
      previousStep: null,
      touchedCube: null,
      state: puzzle.initialState.slice(),
    },
  ];

  while (true) {
    // Take from the beginning of the queue
    let unsolved = workQueue.shift();
    if (unsolved === undefined) break;

    // Now apply every possible state transition
    for (let i = 0; i < puzzle.stateTransitions.length; i++) {
      const newState = unsolved.state.slice();

      // Apply the state transition
      puzzle.stateTransitions[i].forEach((affectedCube) => {
        newState[affectedCube] = mod(newState[affectedCube] + 1, puzzle.maximumNumber);
      });

      const newWork: PuzzleSolveStep = {
        previousStep: unsolved,
        touchedCube: i,
        state: newState,
      };

      // Check if it's a solution
      const isFinalState = newState.every((v, i) => puzzle.finalState[i] == v);
      if (isFinalState) {
        return newWork;
      }

      // Otherwise insert the new state as "work to do"
      if (!solvedStates.has(stateToString(newState))) {
        solvedStates.add(stateToString(newState));
        workQueue.push(newWork);
      }
    }
  }

  return null;
}

function stateToString(state: PuzzleState) {
  return state.map((v) => v + "").join(",");
}

/**
 * Proper modulo, always returns a positive number
 */
function mod(a: number, b: number) {
  return ((a % b) + b) % b;
}
```

**Comment by mgatelabs** [comment]
I think I need to flesh out a Worker that would accept input.  Think i'll have a cube class, direction enum, and the connections.  Then after the worker is finished, just need a way to get the value.  I've never actually used a Worker before.

**Comment by stefnotch** [comment]
> I think I need to flesh out a Worker that would accept input. Think i'll have a cube class, direction enum, and the connections. Then after the worker is finished, just need a way to get the value. I've never actually used a Worker before.

A worker is an absolute overkill for this. Given a crazy large puzzle with 10 stones and every stone has 4 rotations, that would still *just* be a maximum search space of 4*4*4*4... = 4^10 = 1048576 states.
Which means that any halfway decent algorithm should be able to absolutely crush a puzzle in less then a second.

**Comment by mgatelabs** [comment]
I'm looking at the code sample and I think it would need some changes.  There are not a lot of times when it needs a specific direction, they just all need to face the same way or face the same direction as one that cannot move.  I think there is a series of 3 puzzles where some face different ways.  Maybe it needs different solution possabilities, like UNIFIED, they all face the same way, FORCED, they must all face this way, or CUSTOM, where you pass in the array of desired output.

**Comment by stefnotch** [comment]
I suppose one quick tweak would be changing it to an array of `finalStates: PuzzleState[];`. That way, multiple different solutions could be set.

In other news, I found a puzzle that I hadn't completed yet. Here is a shoddy, annotated screenshot. 
- Numbers of cubes
- Which cubes they affect
- Minimap
- Solution (that I found)

![image](https://user-images.githubusercontent.com/10220080/155609295-ca2f6d14-5ed4-4760-8a35-ea7d7ddbe278.png)

**Comment by stefnotch** [comment]
An alternative would be passing a function that checks if a given state is a solution. Here is what a call would look like

```ts
solvePuzzle({
  initialState: [0, 2, 0, 0],
  isFinalState: unifiedSolution,
  // Or one of those two options
  // isFinalState: forcedSolution(0),
  // isFinalState: customSolution([0, 0, 0, 0]),
  maximumNumber: 4,
  stateTransitions: [
    [0,1,3],
    [0,1,2],
    [1,2,3],
    [0,2,3],
  ]
});
```

Tweaked code. Changes are `isFinalState`, refactoring the code to use a helper `addWork` function, and 3 functions at the bottom.
```ts
type PuzzleState = readonly number[];

type Puzzle = {
  /**
   * We represent the state with simple numbers
   */
  initialState: PuzzleState;

  /**
   * Did we reach the final state?
   */
  isFinalState: (state: PuzzleState) => boolean;

  /**
   * What's the maximum state (exclusive), before wrapping around.
   * 4 in the case of a rotating cube puzzle.
   */
  maximumNumber: number;

  /**
   * A state transition basically says "if we hit cube X then Y will be affected"
   */
  stateTransitions: number[][];
};

type PuzzleSolveStep = {
  previousStep: PuzzleSolveStep | null;
  touchedCube: number | null;
  state: PuzzleState;
};

/**
 * @param puzzle The puzzle to be solved
 * @returns A solution if one was found, otherwise null
 */
function solvePuzzle(puzzle: Puzzle) {
  // To prevent doing work more often than we have to
  const solvedStates = new Set<string>();

  // Which puzzle states can we reach and have not solved yet.
  const workQueue: PuzzleSolveStep[] = [];

  function addWork(step: PuzzleSolveStep) {
    const stateString = step.state.map((v) => v + "").join(",");
    if (!solvedStates.has(stateString)) {
      solvedStates.add(stateString);
      workQueue.push(step);
    }
  }

  addWork({
    previousStep: null,
    touchedCube: null,
    state: puzzle.initialState.slice(),
  });

  while (true) {
    // Take from the beginning of the queue
    let unsolved = workQueue.shift();
    if (unsolved === undefined) break;

    // Now apply every possible state transition
    for (let i = 0; i < puzzle.stateTransitions.length; i++) {
      const newState = unsolved.state.slice();

      // Apply the state transition
      puzzle.stateTransitions[i].forEach((affectedCube) => {
        newState[affectedCube] = mod(newState[affectedCube] + 1, puzzle.maximumNumber);
      });

      const newWork: PuzzleSolveStep = {
        previousStep: unsolved,
        touchedCube: i,
        state: newState,
      };

      // Check if it's a solution
      if (puzzle.isFinalState(newState)) {
        return newWork;
      }

      // Otherwise insert the new state as "work to do"
      addWork(newWork);
    }
  }

  return null;
}

/**
 * Proper modulo, always returns a positive number
 */
function mod(a: number, b: number) {
  return ((a % b) + b) % b;
}

function unifiedSolution(state: PuzzleState): boolean {
  return state.every(v => v == state[0]);
}

function forcedSolution(direction: number): (state: PuzzleState) => boolean {
  return (state) => state.every(v => v == direction);
}

function customSolution(finalState: PuzzleState): (state: PuzzleState) => boolean {
  return (state) => state.every((v, i) => finalState[i] == v);
}
```

**Comment by mgatelabs** [comment]
I think we just need the default cube orientation (I think it defaults for each puzzle) and then allow them to click on the cube to rotate it, so it matches what they see, then they could hit solve and get a new solution.

**Comment by mgatelabs** [comment]
So the info from there is:

It is a light puzzle.

initial directions = [2, 3, 3, 1]

0 : connects to 1, 3
1 : connects to 0, 3
2 : connects to 1
3 : connects to 0

**Comment by mgatelabs** [comment]
Derpy Artwork

![Screenshot 2022-02-24 200340](https://user-images.githubusercontent.com/6304749/155639856-e9e6fd22-10c0-4d8e-8f74-c4459a04f13d.png)

**Comment by mgatelabs** [comment]
Cube has been rendered in 3d, and the textures are different on each side

![Screenshot 2022-02-24 214914](https://user-images.githubusercontent.com/6304749/155650229-d7658816-2441-4593-a093-5b73250ba400.png)

**Comment by mgatelabs** [comment]
Getting fancy

Moving around the camera, Window stays at 1.333 aspect ratio.  This way after the image is added, it will more them likely always line up.

![Screenshot 2022-02-24 223318](https://user-images.githubusercontent.com/6304749/155654256-f8937354-2f8d-4aa2-b9f1-cef91c061fa4.png)

**Comment by mgatelabs** [comment]
I guess the 1st puzzle to implement will be the one @stefnotch gave details for.  And it seems like that may have been the last puzzle added to the game, since it's ID is the largest.

I need to re-grab the image, remove the character, maybe use the blur option since we want our cubes on top.  And then crop / make it smaller and try to place it behind the cubes.

I also need to re-work the puzzle component, the puzzle display window need to be it's own component to move all the THREE JS stuff to it's own place.

**Comment by mgatelabs** [comment]
This is sort of getting somewhere

![2022-02-25_13-59-08](https://user-images.githubusercontent.com/6304749/155786748-641b0dbf-e5c7-4e52-8e77-f867c9145f69.png)

**Comment by stefnotch** [comment]
This does look very nice! I can honestly say that I'd enjoy using that tool :)
So, I'm guessing the next steps are 
- Setting the transform (translation and orientation) of the cubes
- Making them click-able

**Comment by mgatelabs** [comment]
Need to change the SPINNING puzzle top so it has an arrow facing forward.

Thinking to start make them match the image (1 area)

- Camera Transform (X, Y, Z), looks at (0,0,0)
- World Rotation (Degrees)
- Cube X / Y / Z (Extra Rotation)

Need methods to change all these values and have it update the camera views.

Then making them clickable.  Need it to emit a message about cube x clicked

Need to make a method to pass in the "configuration".  So we can change the cube rotations.

**Comment by stefnotch** [comment]
I took a look at [making them clickable](https://stackoverflow.com/questions/17638933/three-js-clickable-objects) and realized that the recommended library ([three.interaction.js](https://github.com/jasonChen1982/three.interaction.js)) isn't as well maintained as it should be. 

After some searching, I found [THREE.Interactive](https://github.com/markuslerner/THREE.Interactive). It seems to be a solid alternative, but doesn't have Typescript types! So I did what I could and opened a pull request that fixes [that and a bit](https://github.com/markuslerner/THREE.Interactive/pull/6). Hopefully I didn't break anything there. Let's see what the maintainer says :)

**Comment by mgatelabs** [comment]
Worst case I can add a 2d click layer on top and make the 3d view just output.

**Comment by stefnotch** [comment]
> So the only major actions are like "solve", "reset", "issues". Issues would redirect to the GitHub issues page.

I think we can get rid of "solve", if we just always auto-solve it. Maybe add in some debouncing so that it doesn't go insane after every click.

> Would also need a comment/description block that would have some vague info. "Puzzle is only accessible after using the feather".

That is an excellent idea indeed!

**Comment by mgatelabs** [comment]
If it is fast enough, we can keep a history of previous solutions and if the same one shows up twice, just use the history.

**Comment by mgatelabs** [comment]
Ok, things are getting serious

![Uploading 2022-02-26_21-05-14.gif…]()

**Comment by mgatelabs** [comment]
Adding a selection node.

![2022-02-27_12-09-42](https://user-images.githubusercontent.com/6304749/155894333-63e723d5-7d09-420e-9738-2a7ec55263d0.gif)

**Comment by stefnotch** [comment]
Woah, that's excellent! As for enforcing the same width, either set the width with CSS, or use a monospaced font, or use a table.

**Comment by mgatelabs** [comment]
I started to integrate the puzzle solver code.  Seem to have it working locally, now I just need to reverse the state into a simple array and have it display.

**Comment by mgatelabs** [comment]
Figured out the solver so now I have something like this:

![2022-02-27_17-24-35](https://user-images.githubusercontent.com/6304749/155904436-b4930925-171e-46a4-89d5-96c623a7a9b1.jpg)

Need to figure out the actions, since if you need to show which cube to hit, and maybe you want to see the hit with the previous cube configuration.  Or they would see hit X and show the result.

So would think I would need to add and extra finished step which just shows the end result.

**Comment by mgatelabs** [comment]
Ok, solution now skips the initial step and have the finish step.

![2022-02-27_18-56-02](https://user-images.githubusercontent.com/6304749/155908097-f6abb395-9daf-4e03-8663-e23bec3b85fc.gif)

**Comment by stefnotch** [comment]
![image](https://user-images.githubusercontent.com/10220080/155933828-ee0fecb1-5c07-4dad-8439-f1ade7f5567f.png)

Any idea what's causing this error? It honestly has me stumped, but then again, I'm not used to Angular's less-than-ideal change detection. I prefer [the Vue.js](https://vuejs.org/guide/extras/reactivity-in-depth.html) approach.

**Comment by stefnotch** [comment]
>  Do we have any idea if the back pieces are even hittable?

If the solver doesn't return a solution, then it's guaranteed that there is no solution. It explores the full search space.
In this case, it is possible to create configurations where the back pieces have been hit and there is a valid solution.

> Thinking to hide the initial configuration in an acridian or something similar.

I think it's fine the way it is. Unless, of course, the cubes are clickable...

**Comment by mgatelabs** [comment]
> ![image](https://user-images.githubusercontent.com/10220080/155933828-ee0fecb1-5c07-4dad-8439-f1ade7f5567f.png)
> 
> Any idea what's causing this error? It honestly has me stumped, but then again, I'm not used to Angular's less-than-ideal change detection. I prefer [the Vue.js](https://vuejs.org/guide/extras/reactivity-in-depth.html) approach.

I think I angered the angular gods a bit.  I'll figure it out, probably doesn't like how i'm passing things around.

**Comment by stefnotch** [comment]
Interesting observation indeed. [Judging from this video](https://www.youtube.com/watch?v=DwMfn6xPGXQ), there seem to be a lot of puzzles that have something like that. I suppose one would have to play around with the puzzles more to figure out what exactly they mean.

**Comment by mgatelabs** [comment]
I checked the puzzle at https://mgatelabs.github.io/GenshinSolvers/puzzle?id=x14197 and the 3 back cubes have no interactions, so they will have empty connections and should be skipped when checking.

**Comment by stefnotch** [comment]
> I checked the puzzle at https://mgatelabs.github.io/GenshinSolvers/puzzle?id=x14197 and the 3 back cubes have no interactions, so they will have empty connections and should be skipped when checking.

Interesting indeed. Then I'm glad that the algorithm I whipped up *doesn't* assume that hitting a cube will cause it to rotate.

Like, if we had an imaginary 2 cube puzzle where the user can only touch the second cube:
```ts
solvePuzzle({
  initialState: [3, 0],
  isFinalState: customSolution([3, 3]),
  maximumNumber: 4,
  stateTransitions: [
    [], // Nothing happens when hitting cube 0
    [1], // Cube 1 rotates when cube 1 gets hit
  ]
});
```

**Comment by stefnotch** [comment]
In other news, I just found out that there are a few other Genshin Impact puzzle solvers. 
- https://github.com/Neboer/Genshin-Inadzuma-block-puzzle-solver : Throws some serious mathematics at the puzzle, I'm actually not sure if it will give you an *optimal* solution
- https://github.com/AceDroidX/genshin-puzzle : Is web based
- https://github.com/RwbyChan/genshin-electro-puzzle-solver : Is web based and closed source
- https://github.com/gideonapollob/genshin-puzzle-solver : Isn't intended for the general public, has solutions for 4 puzzles

What they all have in common is that the user has to configure the puzzles themselves. I might reach out to some of them, mainly to say: "Look, we've also built a solver. What do you think?"

**Comment by mgatelabs** [comment]
Nice to know we're not the only ones building an overly complex solution to a simple problem.

**Comment by stefnotch** [comment]
> Nice to know we're not the only ones building an overly complex solution to a simple problem.
> 
> But from what i've witnessed, we might be the simplest, but we have to put in all the leg work.

Yes indeed, I was quite amused to see what other people have tried. I personally really enjoyed the mathematical approach, even if it is like using an industrial jackhammer to tighten a screw. 

And yes again, we're definitely building the simplest one. In fact, I'd go so far as to say that we're building the only one that's use-able enough to actually consider using. Most of the time, the

**Comment by mgatelabs** [comment]
But watching that video, there are cubes with the thing on top and the user can hit them.  Maybe by the time they made those, they were like thing on top pretty, screw the rules.

**Comment by mgatelabs** [comment]
Finally implemented the puzzle that was presented before.

https://mgatelabs.github.io/GenshinSolvers/puzzle?id=x19254

**Comment by mgatelabs** [comment]
Started to test the app on iPhone, it crashed the browser, but all the buttons work.  Should have the click magic working soon.
